### PR TITLE
added annotations to ConcurrentHashMap.java from java 8 JDK

### PR DESCRIPTION
**Pull request by kanha95**
I have modified ConcurrentHashMap.java with all necessary annotations which will solve the issue "Nullness checker handling of ConcurrentHashMap" #1142

### CODE REVIEW

**Comment by mernst** [comment]
As part of this pull request, could you also commit a test case that would fail without the annotated library class but passes with the annotated library class?

**Comment by mernst** [comment]
What you have added is not a test.  For example, there is no way to run it, so no way for it to fail or succeed.
You need to add a file somewhere under `checker/tests/`, probably in `checker/tests/nullness/`.

**Comment by kanha95** [comment]
why i got travis CI build failed mike?

**Comment by mernst** [comment]
Can you say what you have done so far to investigate the problem?

File: checker/jdk/nullness/src/java/util/concurrent/ConcurrentHashMap.java [hunk 1]
```diff
@@ -0,0 +1,6313 @@
...
+
+    /**
+     * Key-value entry.  This class is never exported out as a
+     * user-mutable Map.Entry (i.e., one supporting setValue; see
+     * MapEntry below), but can be used for read-only traversals used
+     * in bulk tasks.  Subclasses of Node with a negative hash field
+     * are special, and contain null keys and values (but are never
+     * exported).  Otherwise, keys and vals are never null.
+     */
+    static class Node<K extends @Nullable Object,V extends @Nullable Object> implements Map.Entry<K,V> {
```
Comment (wmdietl) [inline checker/jdk/nullness/src/java/util/concurrent/ConcurrentHashMap.java]: You should use `@NonNull` as the upper bound, to prevent instantiating with anything that is nullable. Same for the enclosing class.

File: checker/tests/nullness/ConcurrentHashMapTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+myMap.put(3, null);
+
+}
+}
+
+
+
+/*
+Output without annotated ConcurrentHashMap.java: Nullness Checker passed!
```
Comment (wmdietl) [inline checker/tests/nullness/ConcurrentHashMapTest.java]: Properly format your example. Add expected error messages, instead of adding this as a comment.
Make sure the test fails without your change and then passes with your change.

File: checker/jdk/nullness/src/java/util/concurrent/ConcurrentHashMap.java (hunk 1 above)
Comment (kanha95) [inline checker/jdk/nullness/src/java/util/concurrent/ConcurrentHashMap.java]: ok no problem. which methods of java 8 are causing the error ? can you help with that ?

**Comment by kanha95** [comment]
public class ConcurrentHashMap<K,V> extends AbstractMap<K,V>
    implements ConcurrentMap<K,V>, Serializable

This above line implements ConcurrentMap interface. I can't find the ConcurrentMap interface inside the checker-framework. I think this also needs to be added up.

File: checker/jdk/nullness/src/java/util/concurrent/ConcurrentHashMap.java (hunk 1 above)
Comment (wmdietlGC) [inline checker/jdk/nullness/src/java/util/concurrent/ConcurrentHashMap.java]: Do you see the error logs? https://travis-ci.org/typetools/checker-framework/jobs/211887563

One issue is simply that test class names need to be unique per directory. It looks like 'Test' is already used. So pick a different name for your test, ideally based on the issue number.

Java 7-specific errors can be seen here: https://travis-ci.org/typetools/checker-framework/jobs/211887565 Package java.util.function doesn't exist on Java 7.

If your test case uses Java 8 features, add `// @below-java8-jdk-skip-test` somewhere in the beginning, as e.g. seen here: https://github.com/typetools/checker-framework/blob/master/checker/tests/nullness/java8/Issue1000.java

**Comment by wmdietlGC** [comment]
Yes, might need to add other classes as well. As pointed out in https://github.com/typetools/checker-framework/issues/1144 there are also concurrent collections other than maps that should be added.

**Comment by mernst** [comment]
The commit message should include "fixes #[ISSUENUMBER]", to make a connection between the pull request and the issue tracker.

**Comment by kanha95** [comment]
done with all tests except the last one. it points out some "The ' characters around the executable and arguments are not part of the command". I googled it out and found that it's just a warning message .it does not affect the normal flow of the progam. Any suggestion? @mernst @wmdietlGC @wmdietl

**Comment by wmdietlGC** [comment]
`  [apply] Improper formatting: /checker-framework/checker/tests/nullness/ConcurrentHashMapTest.java`

You need to properly format the source code.
You should setup your build environment to prevent yourself from even committing code like this.
Try running `ant reformat`.

**Comment by kanha95** [comment]
@wmdietlGC  i am formatting using netbeans default formatter. Still it's not getting accepted. I am new to this. so kindly suggest.

File: checker/tests/nullness/ConcurrentHashMapTest.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+
+import java.util.concurrent.ConcurrentHashMap;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+class Issue1142 {
```
Comment (mernst) [inline checker/tests/nullness/ConcurrentHashMapTest.java]: In Java, the class name and the file name should match.  Please change one of them.

File: checker/tests/nullness/ConcurrentHashMapTest.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (mernst) [inline checker/tests/nullness/ConcurrentHashMapTest.java]: By convention, when a test is about an issue in the issue tracker, we write a comment at the top of the file.  See other test files in the test suite.

File: checker/jdk/nullness/src/java/util/ConcurrentMap.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package java.util;
```
Comment (mernst) [inline checker/jdk/nullness/src/java/util/ConcurrentMap.java]: This is not the correct package for `ConcurrentMap`.  Why did you change it?
I have the same question about `ConcurrentHashMap`.

**Comment by mernst** [comment]
Werner said,
> Try running `ant reformat`.

He gave a very concrete suggestion.  You disregarded his suggestion and said:

> i am formatting using netbeans default formatter. Still it's not getting accepted. I am new to this. so kindly suggest.

I don't know what else to suggest.  When I tried `ant reformat` on your branch, it worked.

**Comment by kanha95** [comment]
@wmdietlGC  can you please help me in using the ant reformat command ? how to setup the build environment for this? i have never done this type of formatting before. so your help will be valuable to me. i read the manual and but got confused.
Thanks in advance.

**Comment by mernst** [comment]
From the command line, type `ant reformat` and then press the ENTER key.

**Comment by kanha95** [comment]
@mernst  got it...thanks...i messed up with few files. so i am closing this pull request. i will be up with a fresh new PR.
